### PR TITLE
Implement a simple API endpoint for initiating harvests

### LINKED ISSUE #1683: Enable remote clients to initiate harvesting
Currently the `HarvestsController`:
1. validates you are an authenticated user
2. validates you submitted the form with a CSRF token

In order to initiate a request from Airflow, we need to be able to authenticate a different way (probably a token) and we need to not validate CSRF tokens (as well as make sure we are allowed through CORS).

One way to do this might be to set up a new `/api` route that remote clients can access for doing things like kicking off a harvest. We should secure it appropriately.

### PR DESCRIPTION
**Pull request by thatbudakguy**
Closes #1683

### CODE REVIEW

File: app/controllers/api/harvests_controller.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module Api
+  # Initiate a harvest via the API (e.g. from Airflow)
+  class HarvestsController < ApiController
+    delegate :t, to: :I18n
+
+    include NdjsonUploadConcern
+
+    rescue_from ActionController::ParameterMissing, with: :no_url
```
Comment (jcoyne) [inline app/controllers/api/harvests_controller.rb:10]: I'm not a huge fan of `rescue_from` because it makes it hard to follow the program flow.  How about wrapping the lines that raise these with `begin ... rescue` instead?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/controllers/concerns/ndjson_upload_concern.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+# Shared controller methods for uploading ndjson files for harvesting
+module NdjsonUploadConcern
```
Comment (jcoyne) [inline app/controllers/concerns/ndjson_upload_concern.rb:4]: Can we make this a class instead?  Then we can pass in the required URL... That makes it easier to see what the dependencies of this code is.

File: app/controllers/api/harvests_controller.rb (hunk 1 above)
Comment (thatbudakguy) [reply app/controllers/api/harvests_controller.rb:10]: This was my attempt to placate rubocop's branch condition rule, haha. I'm OK with using `begin`; will change it

File: app/controllers/concerns/ndjson_upload_concern.rb (hunk 2 above)
Comment (thatbudakguy) [reply app/controllers/concerns/ndjson_upload_concern.rb:4]: sure, I like that pattern. will do

Comment (thatbudakguy) [reply app/controllers/concerns/ndjson_upload_concern.rb:4]: should it still live in `concerns/`?

Comment (jcoyne) [reply app/controllers/concerns/ndjson_upload_concern.rb:4]: No, probably app/services or similar.

File: app/controllers/api/harvests_controller.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  class HarvestsController < ApiController
+    delegate :t, to: :I18n
+
+    def create
+      begin
+        upload = NdjsonUpload.new(params.require(:url))
+        return invalid_file unless upload.valid?
+      rescue ActionController::ParameterMissing
+        return no_url
+      rescue Errno::ENOENT
```
Comment (jcoyne) [inline app/controllers/api/harvests_controller.rb:14]: Can we move a check for `File.exists?` into `upload.valid?` Then it doesn't have to raise an error in this case?

Comment (thatbudakguy) [reply app/controllers/api/harvests_controller.rb:14]: How would we handle the two error cases separately? I think we still want to preserve the distinction between 422 (filepath OK, bad contents) and 400 (filepath not OK)

Comment (jcoyne) [reply app/controllers/api/harvests_controller.rb:14]: Have an `error` method on the service object. Once you determine that it's not valid, then you can expect the error method to return a value which can be decoded into a response.

Comment (thatbudakguy) [reply app/controllers/api/harvests_controller.rb:14]: I don't think I understand this, but happy to pair if you're willing to explain it

**Comment by aaron-collier** [comment]
Just dropping a comment to say that I've tested this in stage and it's working as expected.

**Review by jcoyne — APPROVED** [review]
